text/0000-float-next-up-down.md

```diff
@@ -0,0 +1,333 @@
+exception that the constants `0x7fff_ffff` and `0x8000_0001` are replaced by
+respectively `0x7fff_ffff_ffff_ffff` and `0x8000_0000_0000_0001`. Using
+exclusively integer arithmetic aids stabilization as a `const fn`, reduces
+transfers between floating point and integer registers or execution units (which
+incur penalties on some processors), and avoids issues with denormal values
+potentially flushing to zero during floating point arithmetic operations
+on some platforms.
+
+```rust
+/// Returns the least number greater than `self`.
+///
+/// Let `TINY` be the smallest representable positive `f32`. Then,
+///  - if `self.is_nan()`, this returns `self`;
+///  - if `self` is `NEG_INFINITY`, this returns `-MAX`;
+///  - if `self` is `-TINY`, this returns -0.0;
+///  - if `self` is -0.0 or +0.0, this returns `TINY`;
+///  - if `self` is `MAX` or `INFINITY`, this returns `INFINITY`;
+///  - otherwise the unique least value greater than `self` is returned.
+///
+/// The identity `x.next_up() == -(-x).next_down()` holds for all `x`. When `x`
+/// is finite `x == x.next_up().next_down()` also holds.
+pub const fn next_up(self) -> Self {
+    const TINY_BITS: u32 = 0x1; // Smallest positive f32.
+    const CLEAR_SIGN_MASK: u32 = 0x7fff_ffff;
+
+    let bits = self.to_bits();
+    if self.is_nan() || bits == Self::INFINITY.to_bits() {
+        return self;
+    }
+    
+    let abs = bits & CLEAR_SIGN_MASK;
+    let next_bits = if abs == 0 {
+        TINY_BITS
+    } else if bits == abs {
+        bits + 1
+    } else {
+        bits - 1
+    };
+    Self::from_bits(next_bits)
+}
+
+/// Returns the greatest number less than `self`.
+///
+/// Let `TINY` be the smallest representable positive `f32`. Then,
+///  - if `self.is_nan()`, this returns `self`;
+///  - if `self` is `INFINITY`, this returns `MAX`;
+///  - if `self` is `TINY`, this returns 0.0;
+///  - if `self` is -0.0 or +0.0, this returns `-TINY`;
+///  - if `self` is `-MAX` or `NEG_INFINITY`, this returns `NEG_INFINITY`;
+///  - otherwise the unique greatest value less than `self` is returned.
+///
+/// The identity `x.next_down() == -(-x).next_up()` holds for all `x`. When `x`
+/// is finite `x == x.next_down().next_up()` also holds.
+pub const fn next_down(self) -> Self {
+    const NEG_TINY_BITS: u32 = 0x8000_0001; // Smallest (in magnitude) negative f32.
+    const CLEAR_SIGN_MASK: u32 = 0x7fff_ffff;
+
+    let bits = self.to_bits();
+    if self.is_nan() || bits == Self::NEG_INFINITY.to_bits() {
+        return self;
+    }
+    
+    let abs = bits & CLEAR_SIGN_MASK;
+    let next_bits = if abs == 0 {
+        NEG_TINY_BITS
+    } else if bits == abs {
+        bits - 1
+    } else {
+        bits + 1
+    };
+    Self::from_bits(next_bits)
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Two more functions get added to `f32` and `f64`, which may be considered
+already cluttered by some.
+
+Additionally, there is a minor pitfall regarding signed zero. Repeatedly calling
```

Comment:
This behaviour follows from the definition of the function, doesn't it? The definition says:

> Returns the largest/smallest number more/less than `self`.

Since `0.0` and `-0.0` compare equal, `0.0.next_down() == -0.0` would violate the contract.

Reply:
Yes, if we didn't want this we would need to change the definition of the function, and depart from IEEE 754 compliance. But I still felt I needed to mention this potential pitfall.